GVFS/GVFS.Common/GVFSEnlistmentHealthCalculator.cs

```diff
@@ -0,0 +1,257 @@
+        /// <returns>A string containing the top level directory from the provided path, or GVFSConstants.GitPathSeparator if the path is for an item in the root</returns>
+        private string ParseTopDirectory(string path)
+        {
+            int whackLocation = path.IndexOf(GVFSConstants.GitPathSeparator);
+            if (whackLocation == -1)
+            {
+                return GVFSConstants.GitPathSeparatorString;
+            }
+
+            return path.Substring(0, whackLocation);
+        }
+
+        /// <summary>
+        /// Categorizes a list of paths given as strings by mapping them to the top level directory in their path
+        /// Modifies the directoryTracking dictionary to have an accurate count of the files underneath a top level directory
+        /// </summary>
+        /// <remarks>
+        /// The distinction between files and directories is important --
+        /// If the path to a file doesn't contain a GVFSConstants.GitPathSeparator, then that means it falls within the root
+        /// However if a directory's path doesn't contain a GVFSConstants.GitPathSeparator, it doesn't count towards its own hydration
+        /// </remarks>
+        /// <param name="paths">An enumerable containing paths as strings</param>
+        /// <param name="directoryTracking">A dictionary used to track the number of files per top level directory</param>
+        /// <param name="parentDirectory">Paths will only be categorized if they are descendants of the parentDirectory</param>
+        private int CategorizePaths(IEnumerable<string> paths, Dictionary<string, int> directoryTracking, string parentDirectory)
+        {
+            int count = 0;
+            foreach (string path in paths)
+            {
+                // Only categorize if descendent of the parentDirectory
+                if (path.StartsWith(parentDirectory, StringComparison.OrdinalIgnoreCase))
+                {
+                    count++;
+
+                    // If the path is to the parentDirectory, ignore it to avoid adding string.Empty to the data structures
+                    if (!parentDirectory.Equals(path, StringComparison.OrdinalIgnoreCase))
+                    {
+                        // Trim the path to parent directory
+                        string topDir = this.ParseTopDirectory(this.TrimDirectoryFromPath(path, parentDirectory));
+                        if (!topDir.Equals(GVFSConstants.GitPathSeparatorString))
+                        {
+                            this.IncreaseDictionaryCounterByKey(directoryTracking, topDir);
+                        }
+                    }
+                }
+            }
+
+            return count;
+        }
+
+        /// <summary>
+        /// Trim the relative path to a directory from the front of a specified path which is its child
+        /// </summary>
+        /// <remarks>Precondition: Only even run in the context of TrimDirectoryFromPath(child, parentDirectory)</remarks>

```

Comment:
`Precondition: Only even run in the context of TrimDirectoryFromPath(child, parentDirectory)`

I'm having trouble understanding what this precondition means.  Are you trying to say something like this?:

`Precondition: 'directoryTarget' must be an ancestor of 'path'`

Reply:
Yeah that wording is better. When I wrote it, it was in response to a comment asking if we needed to check if `directoryTarget` was an ancestor of `path` and I decided to stick a comment in to answer that particular question but it doesn't particularly help readability. Inside the issue: https://github.com/microsoft/VFSForGit/issues/1298